python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -230,24 +234,43 @@ def _handle_missing_modules(cls, containers):
                 shortlist,
             )
             for module in reappeared_modules:
-                cls.event.end(module.device, module.netbox, module.id).save()
+                cls.module_event.end(module.device, module.netbox, module.id).save()
+
+    @classmethod
+    def _handle_new_modules(cls, containers):
+        for shadow_module in containers[cls].values():
+            if shadow_module.is_new:
+                shadow_entity = shadow_module.get_entity()
+                if shadow_entity.is_chassis():
+                    entity = shadow_entity.get_existing_model()
+                    cls.chassis_event.notify(
+                        device=entity.device,
+                        netbox=entity.netbox,
+                        subid=entity.id,
+                        alert_type='chassisNew',
+                    ).save()
```

Comment:
I find this a bit odd. A device cannot (or rather, should not) be both a module and a chassis at the same time.

However, there is potential shenanigans ahead here, due to the dual presentation of stacked members in Juniper virtual chassis, which is slated to get more support in the current sprint. Through standard MIBs, a device may be presented as a module, while through Juniper's proprietary MIBs it will simultaneously be presented as a chassis.

Reply:
The Module model heavily implies that a module can share the same Device as a chassis.

https://github.com/Uninett/nav/blob/05c5213713d6d2cc8d01494721c2f7b2cd31d119/python/nav/models/manage.py#L931-L951

In this code it checks if the NetboxEntity that refers to the same Device as the Module is a chassis. If that is not supposed to be possible, then this code is a bit weird.

But if you say so then i shall trust it, and assume in this PR that it is not possible.

Reply:
Kept the check, but moved Chassis alert code to NetboxEntity shadowclass, as the normal logic should be that chassis are handled by NetboxEntity and modules by Module

Reply:
> In this code it checks if the NetboxEntity that refers to the same Device as the Module is a chassis. If that is not supposed to be possible, then this code is a bit weird.

Not sure how you came to that conclusion. **ENTITY-MIB** presents all entities inside a netbox in a hierarchy, which includes stacks, chassis, modules and ports. This function tries to identify the `self` Module in that hierarchy, then travels up the hierarchy to find the nearest ancestor entity that is of the chassis class, returning that to the caller.

In this model, an entity **cannot** have more than one class. However, the model supports self-loops (though they would make no sense) - and an upstream device is free to represent the same serial number as two different entities at different levels in the hierarchy (which is equally senseless, but still possible).